interfaces/builtin/location.go

```diff
@@ -0,0 +1,298 @@
+getsockname
+recvmsg
+sendmsg
+sendto
+`)
+
+var locationPermanentSlotDBus = []byte(`
+<policy user="root">
+    <allow own="com.ubuntu.location.Service"/>
+    <allow own="com.ubuntu.location.Service.Session"/>
+    <allow send_destination="com.ubuntu.location.Service"/>
+    <allow send_destination="com.ubuntu.location.Service.Session"/>
+    <allow send_interface="com.ubuntu.location.Service"/>
+    <allow send_interface="com.ubuntu.location.Service.Session"/>
+</policy>
+`)
+
+var locationConnectedPlugDBus = []byte(`
+<policy context="default">
+    <deny own="com.ubuntu.location.Service"/>
+    <allow send_destination="com.ubuntu.location.Service"/>
+    <allow send_destination="com.ubuntu.location.Service.Session"/>
+    <allow send_interface="com.ubuntu.location.Service"/>
+    <allow send_interface="com.ubuntu.location.Service.Session"/>
+</policy>
+`)
+
+type LocationInterface struct{}
+
+func (iface *LocationInterface) Name() string {
+	return "location"
+}
+
+func (iface *LocationInterface) PermanentPlugSnippet(plug *interfaces.Plug, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityDBus, interfaces.SecurityAppArmor, interfaces.SecuritySecComp, interfaces.SecurityUDev:
+		return nil, nil
+	default:
+		return nil, interfaces.ErrUnknownSecurity
+	}
+}
+
+func (iface *LocationInterface) ConnectedPlugSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityAppArmor:
+		old := []byte("###SLOT_SECURITY_TAGS###")
+		new := slotAppLabelExpr(slot)
+		snippet := bytes.Replace(locationConnectedPlugAppArmor, old, new, -1)
+		return snippet, nil
+	case interfaces.SecurityDBus:
+		return locationConnectedPlugDBus, nil
+	case interfaces.SecuritySecComp:
+		return locationConnectedPlugSecComp, nil
+	case interfaces.SecurityUDev:
+		return nil, nil
+	default:
+		return nil, interfaces.ErrUnknownSecurity
+	}
+}
+
+func (iface *LocationInterface) PermanentSlotSnippet(slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityAppArmor:
+		return locationPermanentSlotAppArmor, nil
+	case interfaces.SecurityDBus:
+		return locationPermanentSlotDBus, nil
+	case interfaces.SecuritySecComp:
+		return locationPermanentSlotSecComp, nil
+	case interfaces.SecurityUDev:
+		return nil, nil
+	default:
+		return nil, interfaces.ErrUnknownSecurity
+	}
+}
+
+func (iface *LocationInterface) ConnectedSlotSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	// FIXME: This doesn't handle multiple plugs
+	case interfaces.SecurityAppArmor:
+		old := []byte("###PLUG_SECURITY_TAGS###")
+		new := plugAppLabelExpr(plug)
+		snippet := bytes.Replace(locationConnectedSlotAppArmor, old, new, -1)
+		return snippet, nil
```

Comment:
Thinking about it, the pseudo-code for the process is as follows:
```
snippets := getAllSnippets(...)
blob := combineSnippets(snippets, "dbus")
```
Is there something we could do in the combine phase to merge such rules together?

Is there a solution that would not require us to combine this on snapd side and would make it somehow magically work on apparmor side?